### DISCUSSION TITLE
Component-Specific Vercel API Functions

### DISCUSSION THREAD

**Discussion by ryanhaticus** [vercel/next.js#30527] [Ideas]
### Describe the feature you'd like to request

This was just something I thought of that I'd like to see. I'm not sure if there's another feature request out there or adaptation of Next.js that already does this, so here is my suggestion.

Component-specific Vercel API Functions would be awesome.

### Describe the solution you'd like

 What I mean by this is in, for example, my `Users.tsx` file, I could do something like this:

```
// Some sort of decorator to differentiate this function as an API function that should be ran on the server.
@API
const doDatabaseMagic = (someCondition: boolean) => {
  // Some database crap here, serverside code, whatever.
  return users;
}

// Now, for my actual component with examples.
const UsersComponent  = () => {
  const [users, setUsers] = useState(doDatabaseMagic());
   // or...
   useEffect(() => {
     setUsers(doDatabaseMagic());
   }, [someDifferentCondition]);
}
export default UsersComponent;
```

I think it'd be cool for the "compiler" to understand what you want here and automatically create an API route based on your code then connect the frontend / backend seamlessly. To be honest, I am not entirely sure how this works when it comes to that sort of stuff, I am just an end user.

### Describe alternatives you've considered

API routes, obviously... GetStaticProps, etc. All of this is fine for their own use cases, but something universal would be awesome. Also, if this sort of functionality was integrated in each component, development would be more efficient and other people that look at your code can find server logic quicker.

**Comment by sicarius97** [comment]
@ryanhaticus This can all already be done with getServerSideProps. You mentioned in your theoretical doDataBaseMagic function that you could put "database crap, serverside code, or whatever" in there and that is precisely what that is meant for. SPecifically for the advantages you mentioned in some ways, such as the convenience of reviewing code all in one place

**Reply by ryanhaticus** [reply]
Yes, this is sort of what I am looking for. But it'd be nice on our end to be able to call some functions even after the page has loaded and been hydrated. In other words, not passing props to the component on load, but afterwards. I hope that made sense?

**Reply by sicarius97** [reply]
@ryanhaticus I guess I don't quite understand what you mean, what kind of functions exactly are you looking to run after page load? Server side functions or client side functions? If server side, what would be the intended purpose? Generally the purpose of getServerSideProps is to server side render the data and component so that they load together. If you would rather call an API on or after page load, a simple React.useEffect function might be more appropriate.

Again this would all be dependent on the intended purpose you are looking for, but I can certainly help when I have a better idea.
